sycl/doc/developer/Debugging.md

```diff
@@ -0,0 +1,167 @@
+however debug builds can be quite large and slow so using the following `CMake`
+options may help:
+
+- `-DLLVM_USE_SPLIT_DWARF=ON`: Use `-gsplit-dwarf`, this splits some of the
+  debug information out of the object files into their own separate files,
+  which reduces the size of the object files the linker has to load (see
+  [DebugFission](https://gcc.gnu.org/wiki/DebugFission)).
+- `-DLLVM_PARALLEL_LINK_JOBS=4`: Reduce the number of link jobs running in
+  parallel, to avoid running out of RAM when linking large debug build objects.
+- `-DLLVM_USE_LINKER=lld`: Use the LLVM linker `lld` instead of the GNU linker
+  `ld` as `lld` is usually faster.
+- `-DLIBCLC_CUSTOM_LLVM_TOOLS_BINARY_DIR=<path/to/release/build/bin>`: Use a
+  separate release build of DPC++ as the compiler for building the bitcode
+  libraries (`libclc` and `libdevice`), it normally uses the compiler from the
+  same build, but debug clang is quite slow so using a separate release clang
+  can be worth it.
+
+## Tracing the SYCL runtime
+
+The SYCL runtime is built on top of the Unified Runtime API, using the following
+environment variable prints out all the calls to the Unified Runtime emitted by
+the SYCL runtime, which can help understand runtime behavior:
+
+- `SYCL_UR_TRACE=`: Enables SYCL runtime tracing
+  - `1` for basic tracing, `2` for UR call tracing, `-1` for everything.
+
+## Debugging the compiler
+
+### Clang Driver
+
+- `-###`: Prints each command emitted by the clang driver during compilation.
+  - Can be used to manually replay a compilation command step-by-step to narrow
+    down where a crash happened.
+
+### Middle-end and back-end
+
+LLVM has a number of ways to debug LLVM IR passes and lower, the following
+options illustrate a few of them:
+
+- `-save-temps`: Dump all compilation intermediary files.
+- `-mllvm -print-after-all`: Dump modules before and after each pass of the compilation pipeline
+  - Often produces a huge amount of data but can be helpful to track down where
+    something is introduced in the IR or assembly.
+  - To reduce the output size `-mllvm -filter-print-funcs=<function name>`
+    can be used to filter which functions in the module are printed.
+- `-mllvm -opt-bisect-limit=<number>`: Find which optimization pass is causing issues.
+  - In cases where a bug only occurs at a certain optimization level this flag
+    can help bisect which optimization pass is introducing the issue. Using
+    it will print every optimization pass being run with a number, passing that
+    number to the flag makes the optimization pipeline stop any pass after that
+    number. This allows for manual bisection of the issue by adjusting the
+    number passed to the flag. Note that in-lining may interfere with this
+    because it changes the number of `FunctionPass` being run.
+- `-mllvm -debug-only=<tag>`: Enable debug output for given LLVM pass or components.
+  - In LLVM it can be defined as follows `#define DEBUG_TYPE "regalloc"`, this
```

Comment:
We should link the official LLVM docs here: https://www.llvm.org/docs/ProgrammersManual.html#the-llvm-debug-macro-and-debug-option